File: problem238/python/solution.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        output: list[int] = [1] * len(nums)

        prefix_product = 1
        for i in range(1,len(nums)):
            prefix_product = prefix_product * nums[i-1]
            output[i] *= prefix_product

        suffix_product = 1
        for i in range(len(nums)-2,-1,-1):
            suffix_product = suffix_product * nums[i+1]
            output[i] *= suffix_product
        return output

File: problem238/java/Solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        int prev = nums[0];
        result[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            result[i] = prev;
            prev = nums[i] * prev;

        }
        prev = nums[nums.length - 1];

        for (int i = nums.length - 2; i >= 0; i--) {
            result[i] = prev * result[i];
            prev = nums[i] * prev;
        }
        return result;
    }
}
